Feature request: Add url param to debug command

santanaluiz opened the pull request:
The goal is to be able to debug a project without having to compile or have to add a truffle-config file

Related to #4295

haltman-at commented on packages/config/test/methods.test.ts:
```diff
@@ -51,6 +51,16 @@ describe("when it can't find a config file", () => {
+  it("doesn't throw when url is provided", () => {
+    const options = { url: "http://localhost:7545" };
+    const config = TruffleConfig.detect(options);
+    const networks = config.networks
```
This is me being really nitpicky, but, missing semicolon here.  Btw, we now have `prettier` and `eslint` on this repo (dunno when you forked it), so if you run `yarn bootstrap` after cloning it to set things up, `prettier` will automatically fix things like this for you. :)

haltman-at commented on packages/core/lib/commands/debug/meta.js:
```diff
@@ -27,6 +27,10 @@ module.exports = {
       type: "boolean",
       default: false
+    },
+    "url": {
```
Since you added this to debugger, I'm assuming you intended it as a debugger-specific thing.  If it's intended as a general feature then of course it'll have to go with the other general options.

haltman-at commented on packages/core/lib/commands/debug/meta.js:
```diff
@@ -27,6 +27,10 @@ module.exports = {
       default: false
+    },
+    "url": {
+      describe: "Forces/set server URL",
```
I think this description could use some work; also we would normally talk about a "provider" in this context rather than a "server".  Perhaps "Use specified URL for provider"?

haltman-at commented on packages/core/lib/commands/debug/meta.js:
```diff
@@ -59,7 +63,11 @@ module.exports = {
           "Forces the debugger to use artifacts even if it detects a problem.  Dangerous; may cause errors."
-      }
+      },
+      {
```
Maybe move this one up so that it displays earlier?  It's more useful than a dangerous option like `--compile-none`; it probably shouldn't be all the way at the end.  Maybe it could go after `--fetch-external`?

haltman-at replied:
Hm, this was marked resolved, but it appears to be in the same place still?

haltman-at commented on packages/core/lib/commands/debug/meta.js:
```diff
@@ -59,7 +63,11 @@ module.exports = {
+      },
+      {
+        option: "--url",
+        description: "Sets the server URL inline and ignore truffle config file"
```
Again, I'd reword this description.  I'll suggest something like "Connects to a specified provider given via URL, ignoring networks in config"

haltman-at reviewed (changes requested):
Thanks for this!  I think it needs a few changes.  The biggest thing is I'm unsure if `Config.detect` is the right place for the code you added; and using a default config as the base will cause some bugs.  But, this is a good starting point, so thank you for putting this in.

One other comment that doesn't attach to any particular line: I notice you didn't anywhere check that the user didn't set both `--network` and `--url`.  If possible, it might be good to detect that somewhere and error out, because it shouldn't be possible to use both of those simultaneously; they don't make sense together, and we wouldn't want people relying on whatever behavior results from combining them.

cds-amal commented on packages/config/src/index.ts:
```diff
@@ -178,6 +178,21 @@ class TruffleConfig {
     const configPath = options.config;
+    const url = options.url;
+
+    if (url) {
```
I don't understand the implications of this being a general feature option, but it absolutely makes sense to be scoped to the debugger CLI as that's the use case specified by #4295, which is an absolutely brilliant one.

santanaluiz replied:
Makes sense. I agree that adding it to the debugger CLI might make more sense.

What about the possible issues with default configs you just mentioned? I am wondering what can cause problem and if we should combine `--url` always with `--compile-none` for example.
Any thoughts here?

haltman-at replied:
I mean I'm pretty the solution here, is, as mentioned, to merge the new `networks` into the existing config, *not* into a default one.  (Although as part of the idea is that this should work without a config file present, I guess it would use a default one if none exists?)  As I also mentioned, this probably doesn't belong in `Config.detect`, but rather in `core/lib/commands/debug`, around the call to `Config.detect`.  (I guess you'd have to catch errors from the `detect` to handle the no-config case?)

> I am wondering what can cause problem

I mean, Solidity versions are one example; I don't know of others *offhand*, but like -- it's far safer to make specific modifications to the existing config, rather than creating a new one that ignores the existing one entirely.

> if we should combine --url always with --compile-none for example.

No, absolutely not.  The `--compile-none` option is explicitly marked as dangerous; it's not something you would ever want turned on by accident just as a result of using `--url`.

santanaluiz replied:
I did quite some changes. It seems to look better now.

1. `url` param is scoped for the `debug` command only
2. `compileNone` is set to `true` by default
3. user can override `network` inline if they want
4. added tests to debug command

eggplantzzz commented on packages/core/lib/commands/debug/inlineConfigNetwork.js:
```diff
@@ -0,0 +1,17 @@
+      network_id: "*"
+    }
+  };
+  configObject.network = options.network ? options.network : "inline_config";
```
Why are we doing this? In this flow, network will never be set before reaching this line. Will this not be the case some time in the future?

santanaluiz replied:
fair point. I believe we can remove that

eggplantzzz commented on packages/core/lib/commands/debug/inlineConfigNetwork.js:
```diff
@@ -0,0 +1,17 @@
+  const Config = require("@truffle/config");
+
+  const configObject = Config.default();
+  configObject.networks = {
```
This is kind of a thing of principle but instead of overwriting the entire networks property, add to it like
```
configObject.networks["inline_config"] = {...
```

santanaluiz replied:
this makes sense, but if we are using config `default()`, unless it is planned to change in the future, it is an empty object, right?

eggplantzzz replied:
You are correct, in this case it will make no functional difference. But in principle it is probably better to just add something to it rather than overwrite the entire thing.

haltman-at replied:
You can also use `Config.merge()` rather than writing to it directly if you want to be safe!  But writing to it directly should be fine here.

eggplantzzz commented on packages/core/lib/commands/debug/run.js:
```diff
@@ -1,14 +1,21 @@
   const { CLIDebugger } = require("../../debug");
 
-  const config = Config.detect(options);
+  let config = null;
```
Assuming this logic is correct and we want this, a slick way to do this would be the following:
```
const config = options.url ? inlineConfigNetwork(options) : Config.detect(options);
```

haltman-at replied:
Also, minor style matter, but if you just want to create the `config` variable, it makes more sense to just write `let config;` rather than `let config = null;`.  Writing `let config = null;` suggests to the reader that the particular value `null` is somehow meaningful, when in fact it will never get used.

haltman-at commented on packages/core/lib/commands/debug/run.js:
```diff
@@ -1,14 +1,21 @@
 
-  const config = Config.detect(options);
+  let config = null;
+  if (options.url) {
```
So, I don't think this is what we want to be doing.  Just because a `url` was supplied, does not mean that an existing config should be wholly ignored.

Rather, like I said before, I think that if `url` is passed and a config exists, we should use it and merge in the new network; and if one doesn't exist, only then should we create a new one from scratch (and set `compileNone` to `true`).

santanaluiz replied:
makes sense. I misunderstood your previous coment

haltman-at commented on packages/core/lib/commands/debug/inlineConfigNetwork.js:
```diff
@@ -0,0 +1,17 @@
+  };
+  configObject.network = options.network ? options.network : "inline_config";
+  configObject.compileNone = true;
+  configObject._ = [];
```
Why is this line here?  Is this necessary?

santanaluiz replied:
removed

haltman-at commented on packages/core/lib/commands/debug/meta.js:
```diff
@@ -59,6 +63,11 @@ module.exports = {
+      {
+        option: "--url",
+        description:
+          "Connects to a specified provider given via URL, ignoring networks in config. NOTE: compileNone is set to true when url is specified"
```
I don't think we should be setting `compileNone` to `true` whenever `url` is specified.  I think we should only be doing it if `url` is specified and there's no config.

santanaluiz replied:
fixed and updated description

haltman-at reviewed (changes requested):
Thanks!  I still think the logic for determining whether to use an existing config or a new config should be changed, though.

Also, I still think something should be added to `run.js` to check that `network` and `url` weren't both passed in as options.

haltman-at commented on packages/core/lib/commands/debug/meta.js:
```diff
@@ -59,6 +63,11 @@ module.exports = {
+      {
+        option: "--url",
+        description:
+          "Connects to a specified provider given via URL, ignoring networks in config. NOTE: compileNone is set to true when a config file is not found AND url is specified"
```
I think you can leave out the note; this is really a behind-the-scenes thing.  The `compileNone` option isn't dangerous in this context like it usually is.  Although hm... probably it should be made clearer that this allows using the debugger outside of a Truffle project.  I think that part is worth making explicit!  So maybe could just have a second sentence, "Allows using the debugger outside of a Truffle project."?

santanaluiz replied:
Added there, let me know what you think

haltman-at commented on packages/core/lib/commands/debug/mergeConfigNetwork.js:
```diff
@@ -0,0 +1,9 @@
+    network_id: "*"
+  };
+  config.network = inlineConfigNetwork;
+  config.merge(options);
```
So, this works, because `config.merge()` does in fact modify `config` (something I'd forgotten until just now... yikes), but for consistency with our other code I think should actually return the modified config.

haltman-at commented on packages/core/lib/commands/debug/run.js:
```diff
@@ -1,14 +1,19 @@
+  const config = loadConfig(options);
+
+  if (options.url) {
+    mergeConfigNetwork(config, options);
```
Follow-on to previous comment, for consistency I'd suggest `config = mergeConfigNetwork(config, options);`.

haltman-at commented on packages/core/lib/commands/debug/loadConfig.js:
```diff
@@ -0,0 +1,17 @@
+  try {
+    config = Config.detect(options);
+  } catch (error) {
+    if (options.url) {
```
This is maybe being nitpicky, I think it would be good here to also check that the error is of the correct type (like, `error instanceof TruffleError`... maybe check the message too if you want to be really sure, although maybe that's overkill) before invoking this handling, and rethrowing otherwise.

santanaluiz replied:
I wouldn't add the check for the message. If the message is changed in the `Config`, we have a non-functional code here

santanaluiz replied:
added the `instanceof` check, but as I said, I don't think a check for the message is a good idea. Other 2 options would be: 
1. having a constant with the message in Config file - which doesnt look like a good idea
2. Having a more specific error `TruffleConfigFileNotFoundError`, which doesn't seem to be the direction the rest of the codebase is going

Anyhow, seems fine now

santanaluiz replied:
added the `instanceof` check, but as I said, I don't think a check for the message is a good idea. Other 2 options would be: 
1. having a constant with the message in Config file - which doesnt look like a good idea
2. Having a more specific error `TruffleConfigFileNotFoundError`, which doesn't seem to be the direction the rest of the codebase is going

Anyhow, seems fine now

haltman-at replied:
Yeah, that makes sense.  Thanks!

haltman-at commented on packages/core/lib/commands/debug/run.js:
```diff
@@ -1,14 +1,19 @@
   const { CLIDebugger } = require("../../debug");
 
-  const config = Config.detect(options);
+  const config = loadConfig(options);
```
I still think there should be something prior to this that checks that you didn't set both `--network` and `--url`.

santanaluiz replied:
added a check. Not very happy with the copy/wording though. Let me know if you have a better idea :)

haltman-at reviewed (commented):
OK, this is better, thanks again!  A few comments still but they're pretty small.  Some of this is stuff that I commented on earlier, so let me know what you're doing with those... if you intend to leave them alone, you should make a note of that so we all know what's going on.

haltman-at commented on packages/core/test/lib/commands/debug.js:
```diff
@@ -0,0 +1,56 @@
+    });
+
+    it("should create networks item in config", () => {
+      mergeConfigNetwork(config, options);
```
Oh also, for consistency (see previous comments), this should also be `config = ...`;

haltman-at commented on packages/core/lib/commands/debug/meta.js:
```diff
@@ -35,6 +39,11 @@ module.exports = {
+      {
+        option: "--url",
+        description:
+          "Connects to a specified provider given via URL, ignoring networks in config. This option allows using the debugger outside of a Truffle project. That's why compileNone is set to true when a config file is not found AND url is specified"
```
I think the the final sentence here needs to be cut.  That's implementation details that users don't need to know about and will just confuse them; also, it makes this description just too long.

haltman-at commented on packages/core/lib/commands/debug/meta.js:
```diff
@@ -35,6 +39,11 @@ module.exports = {
       OS.EOL +
       "                             [--compile-tests|--compile-all|--compile-none]",
     options: [
+      {
```
The order of these controls the order of display, so I'd suggest moving this down... `--url` shouldn't be the very first thing, the transaction hash should be.  Somewhere after this but before the compile options, I think.

haltman-at reviewed (commented):
Great, thanks so much!  Just one/two minor final things that you can fix really quickly.

(Now could we get a second reviewer on this so we could get this in? :) )

cds-amal commented on packages/core/lib/commands/debug/loadConfig.js:
```diff
@@ -0,0 +1,18 @@
+  try {
+    config = Config.detect(options);
+  } catch (error) {
+    if (error instanceof TruffleError && options.url) {
+      config = Config.default();
+      config.compileNone = true;
```
Can a comment be added to document this error resolution/ code path please?

cds-amal reviewed (changes requested):
Very nice work @santanaluiz ! I left one minor nit about documenting an exception path

santanaluiz commented:
@haltman-at @cds-amal thanks for your reviews. I did the last changes required.
I hope it's fine now :)

cds-amal commented:
Hi @santanaluiz could you update the comments to explain why `compileNone` is set here? It will help future contributors reason about this awesome feature.

santanaluiz commented:
@cds-amal updated the comment, please, share any ideas if you think of a better explanation copy

cds-amal commented:
> @cds-amal updated the comment, please, share any ideas if you think of a better explanation copy

@santanaluiz This update is good. Thank you!

cds-amal reviewed (approved):
I'm excited for this feature @santanaluiz ! Thanks so much. :bow: Here's my [recorded smoke test](https://asciinema.org/a/rc5p0MCS0C07z7aH9QDN1al6R)

santanaluiz commented:
really happy to support 🙇 

Thanks for sharing the smoke test result

haltman-at reviewed (commented):
Sorry, one final thing!  On actually testing this out, I noticed a bug.  If you pass a URL and there's no config, then `compileNone` is supposed to be set.  In reality, this doesn't happen.  This was confusing, but here's why.

Suppose you're in this case. Then `loadConfig` will set `config.compileNone` to true.  However, since `options.compileNone` is still false, the later call to `mergeConfigNetwork` will set `config.compileNone` back to false.

So, uh, something needs to be altered to fix that.  Sorry about that!  But glad I actually tested this. :)

santanaluiz commented:
@haltman-at very good catch! I override this on the options, added a comment for clarity and updated tests

haltman-at commented on packages/core/lib/commands/debug/loadConfig.js:
```diff
@@ -0,0 +1,24 @@
+      config = Config.default();
+      config.compileNone = true;
+      // forcing options to also set compileNone to true, otherwise it can be later override on Config.merge(...)
+      options.compileNone = true;
```
Oh, hm, you fixed it by mutating the `options` that was passed in... I don't think this is a good idea, it means that `loadConfig` has a hidden side effect.  (I know, some of these functions have the hidden side effect of altering `config`, but that we just have to accept.)  Can you fix it without mutating the `options` that was passed in?  A quick way if you want to keep the code this way would be to clone and have this return `{ config, options }` instead of just `config`, maybe not the best way, but just throwing things out...

santanaluiz replied:
actually, I had done some changes meanwhile.
I agree it doesn't look good.
Also, I am not happing returning `config` and `options`

So I did a couple of changes
1. Remove the `mergeConfigNetwork` from `run.js`. Is there a use-case we want to manually add a provider but we can't set `--network` as an argument?
2. Added `mergeConfigNetwork` here to `loadConfig`. So we merge and `options` to defaul config and then set `compileNone` to `true`

wdyt?

haltman-at replied:
Ah!  Yes, I agree that is a better structure.  However there's no `mergeConfigNetwork` in that first branch, there should be one there too, right?

santanaluiz replied:
What do you mean in the first branch? do you mean on the `try` block? 
Why should we have it there? if `url` is provided and we have a config file, we basically ignore `url` anyway, right?

To be honest it might make sense to log a warning, saying the parameter is ignored. But other than that, I don't see a reason. Did I miss something?

haltman-at replied:
Yes, I mean in the `try` block.  My understanding that the desired behavior was that `url` is not ignored in such case, just the network is overridden.

santanaluiz replied:
I see what you mean. Updating...

santanaluiz replied:
should be fine now. Please, let me know of any further point

haltman-at reviewed (approved):
Alright, let's finally get this in! :)  (Assuming nobody else spots any other problems real quick.)

haltman-at reviewed (changes requested):
Blargh, sorry, I'm afraid there is another problem after all.  On actually trying this out, I'm hitting a crash in `getCompilations` in `core/lib/debug/cli.js`.  The problem is that `artifacts` is `undefined` (because the build directory doesn't exist), but since `compileNone` is set, we try to apply `shimArtifacts` to it.  Note we can't just always replace undefined `artifacts` by `[]`, because this mean different things.  But we can do it in specific contexts.

So, we could, for instance, have `gatherArtifacts` return `[]` instead of `undefined` if `compileNone` is set; or we could have `getCompilations` specially handle the possibility of `artifacts` being undefined and `compileNone` being set.  Former seems probably easier to me?  But right now it crashes so I'm afraid we can't merge this with that outstanding. :-/

santanaluiz commented:
Good catch! But I we should check also if config file was skipped. 

Because in the current implementation we can run `$ truffle  debug --url http://host:port --compile-none` 
and it still works in case the user has compiled sources in the `build` directory.

Therefore... 
1. I have added a `skippedConfigFile` field to `config` in this specific use case.
2. Updated tests
3. Added a comment for clarity in there

haltman-at commented:
Oh, you're absolutely right!  This whole time I was conflating `compileNone` with not checking the artifacts... even though that's not what `compileNone` does.  You're right, that should be separate!  Silly me.  OK, will review this now quickly.

haltman-at reviewed (approved):
Great!  Thank you so much!  This all seems to be good now.  Will merge shortly unless anyone has any objections for some reason!
